include/pybind11/detail/internals.h

```diff
@@ -114,7 +114,7 @@ inline void tls_replace_value(PYBIND11_TLS_KEY_REF key, void *value) {
 // libstdc++, this doesn't happen: equality and the type_index hash are based on the type name,
 // which works.  If not under a known-good stl, provide our own name-based hash and equality
 // functions that use the type name.
-#if defined(__GLIBCXX__)
+#if !defined(__apple_build_version__)
```

Comment:
What about non_apple_clang? I am actually not sure we test that in our CI surprisingly, but llvm supports Linux, Windows, and MacOS.

Reply:
Good question.
* Windows clang: I'm trying internal testing, which in the past at least triggered some Windows clang testing, although I never fully understood what exactly.
* Non-Apple macOS clang: Could someone please help with testing?

But:
* I'm not in a rush to merge this PR.
* I just want this to be in the pipeline if and when we decide to increment `PYBIND11_INTERNALS_VERSION`.

FWIW, initially I thought the problem is elsewhere and it may be an easy fix. I didn't anticipate that it would go in the direction you see now.

Reply:
@rwgk Is there an issue opened up on the LLVM repo for this bug btw?

Reply:
> @rwgk Is there an issue opened up on the LLVM repo for this bug btw?

Not to my knowledge.

I think it'll be a couple hours work (at least) to create a minimal reproducer for them, assuming that they will not be able to work with a pybind11-based reproducer.

Reply:
> What about non_apple_clang Clang15, Clang16? I am actually not sure we test that in our CI surprisingly, but llvm supports Linux, Windows, and MacOS.

https://github.com/pybind/pybind11/pull/4316#issuecomment-1310444102

That platform is actually the worst: `std::type_index` does not work as desired, and cross-module exception translation is broken, even if the exceptions have default visibility (IOW "are exported").

Reply:
> Non-Apple macOS clang: Could someone please help with testing?

We could test this using the conda-forge compilers (they are vanilla clang on macOS).

Reply:
>> Non-Apple macOS clang: Could someone please help with testing?
>
> We could test this using the conda-forge compilers (they are vanilla clang on macOS). This also work in CI.

In the meantime I solved this (but forgot to note that here):

https://github.com/pybind/pybind11/pull/4326